File: WebsocketServer/WebsocketServer/DataService/Utiities/TreeItem.cs
using Newtonsoft.Json.Linq;
using WebSocketServer.Utilities;

namespace WebSocketServer.DataService.Utiities
{
    /// <summary>
    /// 可以携带信息的树状数据结构,一对多结构
    /// </summary>
    /// <typeparam name="T"></typeparam>
    public class TreeItem<T>
    {
        public T data;
        /// <summary>
        /// 不允许外部创建，只能继承类和 Collection 创建
        /// </summary>
        protected TreeItem(T data)
        {
            this.data = data;
        }

        /// <summary>
        /// Id 由 Collection 创建，在 Collection 保持唯一
        /// </summary>
        public int Id { get; private set; }
        public int? ParentId { get; private set; }
        public List<int> ChildrenIds { get; private set; } = new List<int>();
       
        /// <summary>
        /// 负责管理树状数据结构
        /// </summary>
        public class Collection
        {
            /// <summary>
            /// flat list tree item，用于网络传输与检索
            /// </summary>
            private readonly List<TreeItem<T>> items = new List<TreeItem<T>>();

            /// <summary>
            /// 获取指定 ID 的节点
            /// </summary>
            /// <param name="id">节点 ID</param>
            /// <returns>节点</returns>
            public TreeItem<T>? GetItemById(int id)
            {
                if (id >= 0 && id < items.Count)
                {
                    return items[id];
                }
                return null;
            }

            /// <summary>
            /// 获取所有节点
            /// </summary>
            /// <returns>所有节点</returns>
            public IEnumerable<TreeItem<T>?> GetAllItems()
            {
                return items;
            }

            /// <summary>
            /// Safe remove item, if the item has any children, it means it's a dependency node, it can not be removed until the children is clear
            /// </summary>
            /// <param name="id">The ID of the item to remove</param>
            /// <returns>True if the item was removed successfully, otherwise false</returns>
            public bool SafeRemoveItem(int id)
            {
                TreeItem<T>? itemToRemove = items.FirstOrDefault(item => item.Id == id);
                if (itemToRemove == null)
                {
                    return false; // Item with the given ID not found
                }

                if (itemToRemove.ChildrenIds.Count > 0)
                {
                    return false; // Item has children, cannot be removed
                }
                return RemoveItem(id);
            }

            /// <summary>
            /// Remove an item from the collection by its ID
            /// </summary>
            /// <param name="id">The ID of the item to remove</param>
            /// <returns>True if the item was removed successfully, otherwise false</returns>
            public bool RemoveItem(int id)
            {
                TreeItem<T>? itemToRemove = items.FirstOrDefault(item => item.Id == id);
                if (itemToRemove != null)
                {
                    int removedIndex = items.IndexOf(itemToRemove);

                    // Remove the item from its parent's ChildrenIds list
                    if (itemToRemove.ParentId.HasValue)
                    {
                        TreeItem<T>? parent = items.FirstOrDefault(item => item.Id == itemToRemove.ParentId);
                        if (parent != null)
                        {
                            parent.ChildrenIds.Remove(id);
                        }
                    }

                    // Remove the item from the flat list
                    items.Remove(itemToRemove);

                    // Adjust IDs of items after the removed item
                    for (int i = removedIndex; i < items.Count; i++)
                    {
                        items[i].Id--;
                    }

                    return true;
                }
                return false;
            }

            /// <summary>
            /// 创建根节点
            /// </summary>
            /// <returns>新节点</returns>
            public TreeItem<T>? CreateRootItem(T data)
            {
                return CreateItem(data, null);
            }

            /// <summary>
            /// 创建新的树状数据结构节点
            /// </summary>
            /// <returns>新节点</returns>
            public TreeItem<T> CreateItem(T data, TreeItem<T>? parent)
            {
                var item = new TreeItem<T>(data)
                {
                    ParentId = parent?.Id,
                    Id = items.Count
                };

                if (parent != null)
                {
                    parent.ChildrenIds.Add(item.Id);
                }

                items.Add(item);
                return item;
            }

            /// <summary>
            /// 从根节点开始操作，parent = null
            /// </summary>
            /// <returns></returns>
            public Result StartFromRoot()
            {
                return new Result(null, null, this);
            }

            /// <summary>
            /// 从某个节点开始操作
            /// </summary>
            /// <returns></returns>
            public Result StartFrom(TreeItem<T>? item)
            {
                return new Result(null, item, this);
            }           
        }

        public class TopmostResult
        {
            private Collection collection;

            public TopmostResult(Collection collection)
            {
                this.collection = collection;
            }

            /// <summary>
            /// 设置没有 parent 的顶层节点
            /// </summary>
            public Result Next(T data)
            {
                var topMostItem = collection.CreateItem(data, null);
                return new Result(topMostItem, null, collection);
            }
        }

        public class Result
        {
            public TreeItem<T>? parent { get; private set; }
            public TreeItem<T>? item { get; private set; }

            private Collection collection;

            public Result(TreeItem<T>? item, TreeItem<T>? parent, Collection collection)
            {
                this.item = item;
                this.parent = parent;
                this.collection = collection;
            }

            /// <summary>
            /// 设置子节点
            /// </summary>
            public Result Children(Action<Result> action)
            {
                action?.Invoke(collection.StartFrom(item));
                return this;
            }

            /// <summary>
            /// 设置平级节点
            /// </summary>
            public Result Next(T data)
            {
                var nextItem = collection.CreateItem(data, parent);
                return new Result(nextItem, parent, collection);
            }
        }
    }
}

File: WebsocketServer/WebsocketServer/wwwroot/TypeScriptClient/lib/utilities/TreeItem.js
var WebsocketTSClient;
(function (WebsocketTSClient) {
    /**
     * 可以携带信息的树状数据结构，一对多结构
     */
    class TreeItem {
        constructor(data) {
            /**
             * 携带的信息
             */
            this.data = null;
            /**
             * Id 由 Collection 创建，在 Collection 保持唯一
             */
            this.id = 0;
            this.parentId = null;
            this.childrenIds = [];
            this.data = data;
        }
        /**
         * to json for network transport, 携带的信息继承了 ITreeItem，也可以一并传输
         */
        toJson() {
            const jobj = {
                Id: this.id,
                ParentId: this.parentId,
                ChildrenIds: this.childrenIds,
                data: this.data
            };
            return jobj;
        }
    }
    WebsocketTSClient.TreeItem = TreeItem;
    (function (TreeItem) {
        /**
        * 负责管理树状数据结构
        */
        class Collection {
            constructor() {
                /**
                 * flat list tree item，用于网络传输与检索
                 */
                this.items = [];
                this.topMostItems = [];
            }
            /**
             * 获取指定 ID 的节点
             */
            getItemById(id) {
                if (id >= 0 && id < this.items.length) {
                    return this.items[id];
                }
                return null;
            }
            getChildrenItemsById(id) {
                const parentItem = this.getItemById(id);
                if (parentItem) {
                    const childrenItems = [];
                    parentItem.childrenIds.forEach(childId => {
                        const childItem = this.getItemById(childId);
                        if (childItem) {
                            childrenItems.push(childItem);
                        }
                    });
                    return childrenItems;
                }
                return null;
            }
            /**
             * 获取所有节点
             */
            getAllItems() {
                return this.items;
            }
            /**
             * 获取所有顶级节点
             */
            getTopMostItems() {
                return this.topMostItems;
            }
            /**
             * Safe remove item, if the item has any children, it means it's a dependency node, it can not be removed until the children are clear
             */
            safeRemoveItem(id) {
                const itemToRemove = this.items.find(item => item.id === id);
                if (!itemToRemove) {
                    return false; // Item with the given ID not found
                }
                if (itemToRemove.childrenIds.length > 0) {
                    return false; // Item has children, cannot be removed
                }
                return this.removeItem(id);
            }
            /**
             * Remove an item from the collection by its ID
             */
            removeItem(id) {
                const itemToRemove = this.items.find(item => item.id === id);
                if (itemToRemove) {
                    if (itemToRemove.parentId === null) {
                        this.topMostItems = this.topMostItems.filter(item => item.id !== id);
                    }
                    const removedIndex = this.items.indexOf(itemToRemove);
                    // Remove the item from its parent's ChildrenIds list
                    if (itemToRemove.parentId !== null) {
                        const parent = this.items.find(item => item.id === itemToRemove.parentId);
                        if (parent) {
                            parent.childrenIds = parent.childrenIds.filter(childId => childId !== id);
                        }
                    }
                    // Remove the item from the flat list
                    this.items.splice(removedIndex, 1);
                    // Adjust IDs of items after the removed item
                    for (let i = removedIndex; i < this.items.length; i++) {
                        this.items[i].id--;
                    }
                    return true;
                }
                return false;
            }
            /**
             * 创建根节点
             */
            createRootItem(data) {
                return this.createItem(data, null);
            }
            /**
             * 创建新的树状数据结构节点
             */
            createItem(data, parent) {
                const item = new TreeItem(data);
                item.parentId = parent ? parent.id : null;
                item.id = this.items.length;
                if (parent) {
                    parent.childrenIds.push(item.id);
                }
                else {
                    // 没有 parent 就是顶级节点
                    this.topMostItems.push(item);
                }
                this.items.push(item);
                return item;
            }
            /**
             * 从根节点开始操作，parent = null
             */
            startFromRoot() {
                return new Result(null, null, this);
            }
            /**
             * 从某个节点开始操作
             */
            startFrom(item) {
                return new Result(null, item, this);
            }
            toJson() {
                const arr = [];
                this.items.forEach(item => {
                    arr.push(item.toJson());
                });
                return arr;
            }
            static parse(json) {
                const collection = new TreeItem.Collection();
                if (Array.isArray(json)) {
                    json.forEach(str => {
                        const itemToken = JSON.parse(str);
                        const id = itemToken['Id'];
                        const parentId = itemToken['ParentId'];
                        const childrenIds = itemToken['ChildrenIds'] || [];
                        const data = itemToken['data'];
                        const treeItem = new TreeItem(data);
                        treeItem.id = id;
                        treeItem.parentId = parentId;
                        treeItem.childrenIds = childrenIds;
                        collection.items.push(treeItem);
                        if (parentId == null) {
                            collection.topMostItems.push(treeItem);
                        }
                    });
                }
                return collection;
            }
        }
        TreeItem.Collection = Collection;
        ;
        class TopmostResult {
            constructor(collection) {
                this.collection = collection;
            }
            /**
             * 设置没有 parent 的顶层节点
             */
            next(data) {
                const topMostItem = this.collection.createItem(data, null);
                return new Result(topMostItem, null, this.collection);
            }
        }
        TreeItem.TopmostResult = TopmostResult;
        ;
        class Result {
            constructor(item, parent, collection) {
                this.parent = null;
                this.item = null;
                this.item = item;
                this.parent = parent;
                this.collection = collection;
            }
            /**
             * 设置子节点
             */
            children(action) {
                action === null || action === void 0 ? void 0 : action(this.collection.startFrom(this.item));
                return this;
            }
            /**
             * 设置平级节点
             */
            next(data) {
                const nextItem = this.collection.createItem(data, this.parent);
                return new Result(nextItem, this.parent, this.collection);
            }
        }
        TreeItem.Result = Result;
        ;
    })(TreeItem = WebsocketTSClient.TreeItem || (WebsocketTSClient.TreeItem = {}));
})(WebsocketTSClient || (WebsocketTSClient = {}));
